### PR TITLE
Various decking fixes

### PR DESCRIPTION
**Pull request by jdevnull**
Just a set of small fixes and messaging clarifications for decking.

1. The maximum number of Hacking Pool dice that can be added to any test is equal to the base number of skill dice in use. (SR3, pg 44).

2. When attempting to logon to an adjacent host, the System Access Node (SAN) may be encrypted, and may be decrypted by targeting the exit's name.  In the mud, this is handled as the destination host's access subsystem being encrypted. When a decker is attempting to pass through the SAN during the midnight hour, in order to prevent the SAN from being repeatedly re-encrypted by matrix_update, we need the destination host to check for the presence of deckers in surrounding hosts.

3. Add messaging for decker dump at 80 security tally. This is in response to https://discord.com/channels/564618629467996170/788953927269613608/1056307563924299989

4.  Fix truncation of matrix download percentage. This is in response to https://discord.com/channels/564618629467996170/788953927269613608/1056394519052034088

5. Clarify whether on the programming or design step when listing software on a computer. This is in response to https://discord.com/channels/564618629467996170/788953927269613608/1056348875335667773

Related to 1, but which I have not been able to figure out, is why GET_MAX_HACKING seems to be running into an unusually low cap every mud hour. See https://discord.com/channels/564618629467996170/1055553152230899792

### CODE REVIEW

File: src/spec_procs.cpp [hunk 1]
```diff
@@ -4375,15 +4375,17 @@ SPECIAL(desktop)
   if (obj->contains) {
     send_to_char(ch, " contains:\r\n");
     for (struct obj_data *soft = obj->contains; soft; soft = soft->next_content) {
-      if (GET_OBJ_TYPE(soft) == ITEM_DESIGN)
```
Comment (luciensadi) [inline src/spec_procs.cpp:4378]: This particular spec has been a pain in my ass for a while, so I just went ahead and rewrote it. Appreciate you doing this though-- someone having to bugfix it yet again prompted me to do the work!

**Review by luciensadi — COMMENTED** [review]
Thanks for the PR! Added a few thoughts.

File: src/newmatrix.cpp [hunk 2]
```diff
@@ -2787,10 +2790,16 @@ void matrix_update()
       if (host.type == HOST_DATASTORE && host.undiscovered_paydata <= 0 && !host.payreset) {
         reset_host_paydata(rnum);
       } else {
-        // See if there are any deckers in here.
-        for (struct matrix_icon *icon = host.icons; icon; icon = icon->next_in_host) {
-          if (!ICON_IS_IC(icon)) {
-            decker = TRUE;
+        // See if there are any deckers here or in surrounding hosts.
```
[the remark this answers is not in the document]
Comment (jdevnull) [reply src/newmatrix.cpp]: Oops!

File: src/newmatrix.cpp [hunk 3]
```diff
@@ -2787,10 +2790,16 @@ void matrix_update()
...
+        // We need to check surrounding hosts to prevent SANs from re-encrypting.
+        for (struct exit_data *exit = host.exit; exit; exit = exit->next) {
+          for (struct matrix_icon *icon = real_host(exit->host).icons; icon; icon = icon->next_in_host) {
+            if (!ICON_IS_IC(icon)) {
+              decker = TRUE;
+              break;
+            }
+          }
+          if (decker) {
```
[the remark this answers is not in the document]
Comment (jdevnull) [reply src/newmatrix.cpp]: Ah, right. Would it make sense to also do this for the inner for loop as well, or would it be more readable leaving the break in there?
```cpp
for (struct matrix_icon *icon = real_host(exit->host).icons; icon; icon = icon->next_in_host) {
  if (!ICON_IS_IC(icon)) {
    decker = TRUE;
    break;
  }
}
```
-- vs --
```cpp
for (struct matrix_icon *icon = real_host(exit->host).icons; icon && !decker; icon = icon->next_in_host) {
  if (!ICON_IS_IC(icon)) {
    decker = TRUE;
  }
}
```

[the remark this answers is not in the document]
Comment (luciensadi) [reply src/newmatrix.cpp]: The former works for me! I really only change the continuation logic for the outer parts of nested loops.

File: src/spec_procs.cpp (hunk 1 above)
Comment (jdevnull) [reply src/spec_procs.cpp]: I'm good with that :)

My linux machine is my work machine so I can't put an instance of the mud on there, and my laptop is windows, so I'm using the lack of a testing environment as my excuse for not tackling more substantial re-writes.

File: src/newmatrix.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (jdevnull) [reply src/newmatrix.cpp]: Got it

**Comment by luciensadi** [comment]
Looks good, thanks!

**Pull request merged**